### PR TITLE
Reworked Ctrl+Tab to work without RepositoryView focus

### LINKED ISSUE #5911: Changes/History do not switch with Ctrl-Tab shortcut if the focus is outside of the changes/history list or content view
## Description

Changes/History do not switch with Ctrl-Tab shortcut if the focus is outside of the changes/history list or content view.

## Version

* GitHub Desktop: 1.4.2

* Operating system:
macOS 10.14

## Steps to Reproduce

1. Click "Fetch ..." button
2. Press Ctrl-Tab -> nothing happened

Alternative steps:
1. Select any view in Changes/History tab or content view
    a. item in the changes/history list
    b. Summary or Description field in the Changes tab
    c. changed files list or diff view in the content view
2. Press Ctrl-Tab -> tabs switch (as expected)
3. Press Ctrl-Tab again -> nothing happens

### Expected Behavior

Pressing Ctrl-Tab should always switch tabs as long as tabs are present on the screen and there is no modal dialog, popup or any other obstructing view is present on the screen

### Actual Behavior

Pressing Ctrl-Tab switches tabs only when the focus is set to a view inside Changes/History tab or content view

### PR DESCRIPTION
**Pull request by JMcKiern**
## Overview

See title.
**Closes #5911**

## Description

Ctrl+Tab allows the user to change the tab in sidebar (from Changes to History or vice-versa).
Previously it required RepositoryView to have the focus.
Now it works if the app has focus.

## Release notes

Notes: no-notes

### CODE REVIEW

File: app/src/ui/repository.tsx [hunk 1]
```diff
@@ -379,24 +380,40 @@ export class RepositoryView extends React.Component<
     this.props.dispatcher.revertCommit(this.props.repository, commit)
   }
 
-  private onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
+  public componentDidMount() {
+    window.addEventListener('keydown', this.onGlobalKeyDown)
```
[the remark this answers is not in the document]
Comment (JMcKiern) [reply app/src/ui/repository.tsx]: Ah ok, didn't realise that. I was basically just mimicking what had been done in [app.tsx](https://github.com/desktop/desktop/blob/387cf9d42da54e7f679458e92dd2a385e95ef5b6/app/src/ui/app.tsx#L839) for the Alt key behaviour on Windows.

File: app/src/ui/app.tsx [hunk 2]
```diff
@@ -2376,6 +2376,7 @@ export class App extends React.Component<IAppProps, IAppState> {
           onOpenInExternalEditor={this.openFileInExternalEditor}
           appMenu={this.state.appMenuState[0]}
           currentTutorialStep={this.state.currentOnboardingTutorialStep}
+          isShowingModal={this.isShowingModal}
```
[the remark this answers is not in the document]
Comment (JMcKiern) [reply app/src/ui/app.tsx]: Feel free to suggest anything better!

File: app/src/ui/repository.tsx [hunk 3]
```diff
@@ -379,24 +380,40 @@ export class RepositoryView extends React.Component<
...
+  }
+
+  private onGlobalKeyDown = (event: KeyboardEvent) => {
+    if (event.defaultPrevented) {
+      return
+    }
+
+    if (this.props.isShowingModal) {
```
[the remark this answers is not in the document]
Comment (JMcKiern) [reply app/src/ui/repository.tsx]: Yup. As before, I was keeping it similar to how it's done in app.tsx: https://github.com/desktop/desktop/blob/387cf9d42da54e7f679458e92dd2a385e95ef5b6/app/src/ui/app.tsx#L852-L859

But I can change it if that would make sense. Do you know if there was some reason to split them up in app.tsx?

**Comment by JMcKiern** [comment]
Thanks!

**Comment by niik** [comment]
Hey @JMcKiern, I'm so sorry for letting this sit for so long. I'd be happy to work with you to get this merged. Are you still interested in working on it?

I did a quick review of the changes and they look great, the only thing I'd like to add is that in addition to checking for `isShowingModal` we should also check for `isShowingFoldout` (that property doesn't exist but it's just checking for `this.state.currentFoldout !== null`. That way we don't switch tabs in the repository view when hitting `Ctrl+Tab` in say the branches dropdown.

We should also add a tiny bit of documentation to the newly introduced `isShowingModal`/`isShowingFoldout` in the `RepositoryView`. I'd steal the documentation from `App.isShowingModal` and tweak it just a little.

> A value indicating whether or not the application is currently presenting a modal dialog such as the preferences, or an error dialog.

I went ahead and resolved the merge conflict for you in case you want to pick this up.

**Comment by JMcKiern** [comment]
Sure, I'll have a go at it now

**Comment by JMcKiern** [comment]
That's ready now. Let me know if I need to change something

**Comment by ruggi99** [comment]
Hi all,
I would like to suggest another way to detect opened dialogs and foldouts.
How about putting `event.stopPropagation()` in keydown event listeners of dialogs and foldouts?
PR #9857 ensures that only dialog or its children are focused, so event listener on dialog collects every keydown event.
This will prevent keydown events from trigger at window level and so deactivate all shortcuts.
Menu shortcut are not affected, if I'n not wrong.

**Pull request merged**